src/nn/tensor_collection/visitor.rs

```diff
@@ -0,0 +1,198 @@
+use crate::{
+    shapes::{Dtype, Shape},
+    tensor::{DeviceStorage, Tensor},
+};
+
+use super::collection::{ModuleVisitor, TensorCollection, TensorOptions};
+
+use std::{string::String, vec::Vec};
+
+/// A standard [ModuleVisitor] that executes `F` on every [Tensor] encountered.
+/// `F` must implement [TensorVisitor]
+pub struct RecursiveWalker<'a, M, F> {
+    pub m: M,
+    pub f: &'a mut F,
+    pub path: &'a mut Vec<String>,
+}
+
+/// Something that can visit [Tensor]s. Used in conjunction with [RecursiveWalker].
+pub trait TensorVisitor<E: Dtype, D: DeviceStorage> {
+    /// The type of tensor this struct uses. E.g. [TensorMut], or [TensorRef]
+    type Viewer: TensorViewer;
+    type Err;
+
+    fn visit<S: Shape>(
+        &mut self,
+        full_path: String,
+        opts: TensorOptions<S, E, D>,
+        t: <Self::Viewer as TensorViewer>::View<'_, Tensor<S, E, D>>,
+    ) -> Result<(), Self::Err>;
+}
+
+/// Something that can view [Tensor]s in different ways. For example
+/// [ViewTensorRef] can view `&Tensor`, and [ViewTensorMut] can view `&mut Tensor.
+pub trait TensorViewer: 'static {
+    type View<'a, Mod: 'a>
+    where
+        Self: 'a;
+
+    /// Return the view of the tensor
+    fn view<'a, Mod, Field, GetRef, GetMut>(
+        module: &'a mut Self::View<'_, Mod>,
+        get_ref: &mut GetRef,
+        get_mut: &mut GetMut,
+    ) -> Self::View<'a, Field>
+    where
+        GetRef: FnMut(&Mod) -> &Field,
+        GetMut: FnMut(&mut Mod) -> &mut Field;
```

Comment:
I like using 'view' to express the meaning of this trait, but the naming/documentation of the view function isn't very informative.
```suggestion
    /// Given a view to a module, returns a view to one of that module's fields
    fn view_field<'a, Mod, Field, GetRef, GetMut>(
        module: &'a mut Self::View<'_, Mod>,
        get_ref: &mut GetRef,
        get_mut: &mut GetMut,
    ) -> Self::View<'a, Field>
    where
        GetRef: FnMut(&Mod) -> &Field,
        GetMut: FnMut(&mut Mod) -> &mut Field;
```